core/src/test/java/org/apache/druid/testing/junit/LoggerCaptureRule.java

```diff
@@ -94,7 +94,7 @@ public void append(LogEvent logEvent)
 
     List<LogEvent> getLogEvents()
     {
-      return Collections.unmodifiableList(logEvents);
+      return Collections.synchronizedList(logEvents);
```

Comment:
1) It's not copy-on-write (as the title of your PR says), it's `synchronizedList`, but this doesn't solve the problem: see https://github.com/code-review-checklists/java-concurrency#synchronized-collection-iter. However, if you actually use `CopyOnWriteArrayList` instead of `ArrayList` as `logEvents`, that would help. Don't forget to change the type of the field from `List<LogEvent>` to `CopyOnWriteArrayList<LogEvent>` and to add a comment explaining the need for synchronization, along the lines of https://github.com/code-review-checklists/java-concurrency#justify-document.